pydicom/tests/test_handler_util.py

```diff
@@ -1362,3 +1367,511 @@ def test_unknown_opcode_raises(self):
+        assert 642 == arr[0, 326, 130]
+        assert 3453 == arr[1, 326, 130]
+        out = apply_voi_lut(arr, ds)
+        assert 3046.6 == pytest.approx(out[0, 326, 130], abs=0.1)
+        assert 4095.0 == pytest.approx(out[1, 326, 130], abs=0.1)
+
+    def test_window_rescale(self):
+        """Test windowing after a rescale operation."""
+        ds = dcmread(WIN_12_1F)
+        assert 16 == ds.BitsAllocated
+        assert 12 == ds.BitsStored
+        assert 0 == ds.PixelRepresentation
+        assert [450, 200] == ds.WindowCenter
+        assert [790, 443] == ds.WindowWidth
+        ds.RescaleSlope = 1.2
+        ds.RescaleIntercept = 0
+
+        arr = ds.pixel_array
+        assert 0 == arr[16, 60]
+        assert 642 == arr[326, 130]
+        assert 1123 == arr[316, 481]
+        hu = apply_modality_lut(arr, ds)
+        assert 0 == hu[16, 60]
+        assert 770.4 == hu[326, 130]
+        assert 1347.6 == hu[316, 481]
+        # With rescale -> output range is 0 to 4914
+        out = apply_voi_lut(hu, ds)
+        assert 0 == pytest.approx(out[16, 60], abs=0.1)
+        assert 4455.6 == pytest.approx(out[326, 130], abs=0.1)
+        assert 4914.0 == pytest.approx(out[316, 481], abs=0.1)
+
+    def test_window_bad_photometric_interp(self):
+        """Test bad photometric interpretation raises exception."""
+        ds = dcmread(WIN_12_1F)
+        ds.PhotometricInterpretation = 'RGB'
+        msg = r"only 'MONOCHROME1' and 'MONOCHROME2' are allowed"
+        with pytest.raises(ValueError, match=msg):
+            apply_voi_lut(ds.pixel_array, ds)
+
+    def test_window_bad_parameters(self):
+        """Test bad windowing parameters raise exceptions."""
+        ds = dcmread(WIN_12_1F)
+        ds.WindowWidth = 0
+        ds.VOILUTFunction = 'LINEAR'
+        msg = r"Width must be greater than or equal to 1"
+        with pytest.raises(ValueError, match=msg):
+            apply_voi_lut(ds.pixel_array, ds)
+
+        ds.VOILUTFunction = 'LINEAR_EXACT'
+        msg = r"Width must be greater than 0"
+        with pytest.raises(ValueError, match=msg):
+            apply_voi_lut(ds.pixel_array, ds)
+
+        ds.VOILUTFunction = 'SIGMOID'
+        msg = r"Width must be greater than 0"
+        with pytest.raises(ValueError, match=msg):
+            apply_voi_lut(ds.pixel_array, ds)
+
+        ds.VOILUTFunction = 'UNKNOWN'
+        msg = r"Unsupported \(0028,1056\) VOI LUT Function value 'UNKNOWN'"
+        with pytest.raises(ValueError, match=msg):
+            apply_voi_lut(ds.pixel_array, ds)
+
+    def test_unchanged(self):
+        """Test input array is unchanged if no VOI LUT"""
+        ds = Dataset()
+        ds.PhotometricInterpretation = 'MONOCHROME1'
+        ds.PixelRepresentation = 1
+        ds.BitsStored = 8
+        arr = np.asarray([-128, -127, -1, 0, 1, 126, 127], dtype='int8')
+        out = apply_voi_lut(arr, ds)
+        assert [-128, -127, -1, 0, 1, 126, 127] == out.tolist()
```

Comment:
As usually, I only skipped through the tests - checking out all values would take too much time, and I'm lazy...